### PR TITLE
General mods and fixes

### PR DESCRIPTION
**Pull request by rlustemberg**
Fixes in Session and SessionConfiguration proxies
Modified uploadTask to use file from disk instead of data
Added method and supplementary ‘headers’  dictionary for full control on upload request (on uploadTask configuration dictionary

### CODE REVIEW

File: ios/Classes/ComAppceleratorUrlSessionModule.m [hunk 1]
```diff
@@ -128,17 +128,34 @@ -(id)addBackgroundUploadTask:(id)args
...
-    TiBlob *blob = nil;
+    NSString *fileUrlString = nil;
+    NSString *method = nil;
+    NSURL *fileURL = nil;
+    NSDictionary *headers = nil;
     
     ENSURE_ARG_FOR_KEY(session, args, @"session", ComAppceleratorUrlSessionSessionProxy);
     ENSURE_ARG_FOR_KEY(url, args, @"url", NSString);
-    ENSURE_ARG_FOR_KEY(blob, args, @"data", TiBlob);
+    ENSURE_ARG_FOR_KEY(fileUrlString, args, @"file", NSString);
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionModule.m]: `file` and `data` are supposed to do the same thing. Because Ti.Network.HTTPClient uses "data", we should make the type `id` and handle the type (`Ti.Blob` or `NSString`) afterwards to support both. But I can do that afterwards.

File: ios/Classes/ComAppceleratorUrlSessionModule.m [hunk 2]
```diff
@@ -128,17 +128,34 @@ -(id)addBackgroundUploadTask:(id)args
...
     
     if ([session session] != nil) {
         if ([url length] != 0) {
-            NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:url]];
-            NSURLSessionUploadTask *task = [[session session] uploadTaskWithRequest:request
-                                                                                          fromData:[blob data]];
+            NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url]];
+            [request setHTTPMethod:method];
+            //additional headers which cannot be set on the sessionConfiguration object
+            if([args objectForKey:@"httpHeaderOverrides"]){
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionModule.m]: I will rename this to `requestHeaders` to conform with other SDK API's

File: ios/Classes/ComAppceleratorUrlSessionModule.m [hunk 3]
```diff
@@ -159,7 +176,6 @@ -(id)backgroundDownloadTaskWithURL:(id)args
     return [self addBackgroundDownloadTask:@{
         @"session": [args objectAtIndex:0],
         @"url": [args objectAtIndex:1],
-        @"data": [args objectAtIndex:2],
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionModule.m:162]: Why was this removed?

File: ios/Classes/ComAppceleratorUrlSessionSessionProxy.m [hunk 4]
```diff
@@ -25,6 +25,15 @@ - (id)_initWithPageContext:(id<TiEvaluator>)context andArguments:(NSDictionary*)
...
     return self;
 }
+-(void)_initWithProperties:(NSDictionary *)properties
+{
+    sessionConfiguration = [properties objectForKey:@"configuration"];
+    [self rememberProxy:sessionConfiguration];
+    session = [NSURLSession sessionWithConfiguration:[sessionConfiguration sessionConfiguration]
+                                            delegate:(id<NSURLSessionDelegate>)[[UIApplication sharedApplication] delegate]
+                                       delegateQueue:nil];
+    [super _initWithProperties:properties];
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionSessionProxy.m:35]: Same here: This whole logic has already been done in `_initWithPageContext:andArguments` above, why would you do it here as well?

File: ios/Classes/ComAppceleratorUrlSessionSessionConfigurationProxy.m [hunk 5]
```diff
@@ -23,9 +24,20 @@ - (id)_initWithPageContext:(id<TiEvaluator>)context andArguments:(NSDictionary*)
...
+{
+    
+    NSString *identifier = [properties objectForKey:@"identifier"];
+    
+    if ([TiUtils isIOS8OrGreater] == YES) {
+        sessionConfiguration = [NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:identifier];
+    } else {
+        sessionConfiguration = [NSURLSessionConfiguration backgroundSessionConfiguration:identifier];
+    }
+    [super _initWithProperties:properties];
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionSessionConfigurationProxy.m:39]: This whole logic has already been done in `_initWithPageContext:andArguments` above, why would you do it here as well?

**Review by hansemannn — CHANGES_REQUESTED** [review]
Left some changes need to be addressed before.

File: ios/Classes/ComAppceleratorUrlSessionModule.m (hunk 1 above)
Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionModule.m]: Sounds good. For background uploads with the app in suspended mode, we do need to use the `NSURLSessionUploadTask *task = [[session session] uploadTaskWithRequest:request fromFile:fileURL];` selector. So I'll handle both cases.

File: ios/Classes/ComAppceleratorUrlSessionModule.m (hunk 2 above)
Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionModule.m]: Perfect. I'll do that on my next commit.

File: ios/Classes/ComAppceleratorUrlSessionModule.m (hunk 3 above)
Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionModule.m:162]: It was not being used in the download task

File: ios/Classes/ComAppceleratorUrlSessionSessionProxy.m (hunk 4 above)
Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionSessionProxy.m:35]: It was needed because the `_initWithPageContext:andArguments` is not being called when you are instantiating the object directly from JavaScript, which was leaving the session instance variable unset. I left `_initWithPageContext:andArguments` to handle the legacy case, when we are initialising from inside the legacy module method. I am not sure if there is another way of doing this, as the properties are handled once the object is already initialised, afaik.

File: ios/Classes/ComAppceleratorUrlSessionSessionConfigurationProxy.m (hunk 5 above)
Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionSessionConfigurationProxy.m:39]: Same as above

File: ios/Classes/ComAppceleratorUrlSessionSessionProxy.m (hunk 4 above)
Comment (hansemannn) [reply ios/Classes/ComAppceleratorUrlSessionSessionProxy.m:35]: The `_initWithPageContext:` should be invoked in [here](https://github.com/rlustemberg/ti.urlsession/blob/30989c2ac0803fd0b96d7f4648c0fa0df53cd163/ios/Classes/ComAppceleratorUrlSessionModule.m#L76) from the JS side. Can you give me a JS example that won't call it?

Comment (rlustemberg) [reply ios/Classes/ComAppceleratorUrlSessionSessionProxy.m:35]: If I use tirulesession.createURLSession(opts) (I call this the legacy method) , it's allright. But if I create the proxy directly by doing tirulesession.createSession(opts), then it doesnt call `_initWithPageContext:`

File: ios/Classes/ComAppceleratorUrlSessionModule.m [hunk 6]
```diff
@@ -127,35 +127,46 @@ -(id)addBackgroundUploadTask:(id)args
...
                 for( id key in headers){
                     ENSURE_TYPE(key, NSString);
                     ENSURE_TYPE([headers objectForKey:key], NSString);
                     [request setValue:[headers objectForKey:key] forHTTPHeaderField:key];
                 }
             }
-            NSURLSessionUploadTask *task = [[session session] uploadTaskWithRequest:request fromFile:fileURL];
+            if([data isKindOfClass:[NSString class]]){
+                fileURL = [NSURL fileURLWithPath:data];
+                task = [[session session] uploadTaskWithRequest:request fromFile:fileURL];
```
Comment (hansemannn) [inline ios/Classes/ComAppceleratorUrlSessionModule.m]: Use `[TiUtils toURL:fileURL]` to handle more url-formats.

File: ios/Classes/ComAppceleratorUrlSessionSessionProxy.m (hunk 4 above)
Comment (hansemannn) [reply ios/Classes/ComAppceleratorUrlSessionSessionProxy.m:35]: Checked it. We should rename `ComAppceleratorUrlSessionSessionProxy` to `ComAppceleratorUrlSessionURLSessionProxy`. We don't deprecate the old one, so people should still use `createURLSession`.

**Comment by hansemannn** [comment]
Merging to my PR for now, going from there! Thanks so much @rlustemberg !

**Review by hansemannn — APPROVED** [review]

**Comment by rlustemberg** [comment]
I'll start working on the PR to the sdk within a couple of hours. Cheers!
